CompareRaftMembers/BucketStream.js

```diff
@@ -116,26 +116,71 @@ class BucketStream extends stream.Readable {
+                return async.mapLimit(Contents, 10, (item, itemCb) => {
+                    if (this.lastKey && item.key > this.lastKey) {
+                        ended = true;
+                        return itemCb(null, null);
+                    }
+                    const metadata = this._preprocessItem(item);
+                    if (!metadata) {
+                        return itemCb(null, null);
+                    }
+                    if (metadata.location !== undefined) {
+                        return itemCb(null, item);
+                    }
+                    // need to fetch the object metadata to retrieve
+                    // the full location array
+                    return this._fetchObjectMetadata(item.key, (err, completeMetadata) => {
+                        if (err) {
+                            return itemCb(err);
+                        }
+                        if (!completeMetadata) {
+                            // can happen if the object was not found,
+                            // just skip it in that case
+                            return itemCb(null, null);
+                        }
+                        return itemCb(null, { key: item.key, value: completeMetadata });
+                    });
+                }, (err, listing) => {
+                    if (err) {
+                        // there was an error fetching metadata after
+                        // retries: destroy the stream
+                        return this.destroy(err);
+                    }
+                    for (const item of listing) {
+                        if (item) {
+                            const fullKey = `${this.bucketName}/${item.key}`;
                             this.push({
                                 key: fullKey,
                                 value: item.value,
                             });
                         }
+                    }
+                    if (IsTruncated && !ended) {
+                        this.marker = Contents[Contents.length - 1].key;
                     } else {
-                        ended = true;
-                        break;
+                        this.push(null);
                     }
+                    return undefined;
+                });
+            },
+        );
+    }
+
+    _fetchObjectMetadata(objectKey, callback) {
+        const url = `/default/bucket/${this.bucketName}/${encodeURIComponent(objectKey)}`;
+        return async.retry(
+            this.retryParams,
+            done => this._requestWrapper({ method: 'GET', path: url }, done),
+            (err, response) => {
+                if (err) {
+                    return callback(err);
                 }
-                if (IsTruncated && !ended) {
-                    this.marker = Contents[Contents.length - 1].key;
-                } else {
-                    this.push(null);
+                if (!response) {
+                    // the object was not found
```

Comment:
maybe adding a log here is useful?

Reply:
I think it's not too useful to log this, as it can be a normal situation depending on circumstances. Moreover, it would require carrying a logger in this case, which is possible, but I'd rather have it decoupled from the logging and emit error events that can be logged by the main script when errors occur.